### DISCUSSION TITLE
Records: prevent object cloning if resulting object is logically the same

### DISCUSSION THREAD

**Discussion by macsux** [dotnet/csharplang#8506] [Language Ideas]
Currently, a call to a records `with` expression always produces a new instance. This however is not optimal if all the new values passed to the `with` clause are the same as the original record. Consider the following:
```csharp
public record Node(string MyProperty);
...
var original = new Node("A");
var derived = original with { MyProperty = "A" };
Console.WriteLine(derived.Equals(original)); // true
Console.WriteLine(derived == original); // true
Console.WriteLine(ReferenceEquals(derived, original)); // false


```

In certain scenarios where happens there's opportunity to reuse existing objects rather than cloning, potentially preventing significant new memory allocations. Since the two objects are immutable, returning original in place of copy should be safe.

This behaviour should probably be on an "opt-in" basis because it would require comparison of the new properties to ones on the existing object before deciding if object needs to be cloned. 

A practical example where this might be useful is in code that implements a visitor pattern over immutable data structure (like Roslyn) where 90% of the tree structure could be reused and only small "branch" of nodes changed.

### Implementation proposal
_I'll update this as discussion below progresses_
This change would require the use of upgraded compiler on both side, hence requiring that libraries that choose to use this feature re-target to newer versions of .NET.

Compiler generates a synthetic method like this on record types:
```csharp
protected virtual bool ShouldClone(string PropertyA, int PropertyB) => true;
```

If overriden, the compiler would instead of generating the normal hidden parameter-less `<Clone>$` method, would instead one that like this:
```csharp
public virtual A <Clone>$(string PropertyA, int PropertyB)
{
    return ShouldClone(PropertyA, PropertyB) ? new A(this) : this;
}
```

Consumer side compiler would interpret a `with` clause like this:
```csharp
var clone = a with 
{
    PropertyA = "a",
    PropertyB = "b"
}
```
into something logically similar to this:
```csharp
var clone = a.<Clone>$("a","b");
clone.PropertyA = "a";
clone.PropertyB = "b";
```

**Comment by HaloFour** [comment]
Records are cloned prior to any of the properties being set:

```cs
var original = new Node("A");

// the following
var derived = original with { MyProperty = "A" };

// is equivalent to
var $temp = original.Clone();
$temp.MyProperty = "A";
var derived = $temp;
```

**Reply by macsux** [reply]
Yes, I'm quite aware of this fact (though not possible to do exactly like you wrote it because MyProperty is `init` only, but resulting IL doesn't care about that). If some kinda of language extension is added, the compiler would instead "logically" rewrite it to something like this:

```csharp
Node derived;
if(original.MyProperty == "A")
{
    derived = original;
}
else
{
    var $temp = original.Clone();
    $temp.MyProperty = "A";
    derived = $temp;
}
```

**Reply by HaloFour** [reply]
This would also be a source breaking change, in that the behavior of an existing program would change when recompiled.  That can be observable if the cloned return is used with `lock` or in a collection with reference identity.

The single allocation is also probably a wash with the work needed to perform this comparison, especially for larger records.

**Reply by macsux** [reply]
It can be breaking, but if it's implemented on an opt-in basis then the authors would have made that as a deliberate choice. 

There are practical reasons to do this beyond just memory allocations. For example, serializers will often make an optimization by adding some kinda `ref` pointer to an object lookup table if the same object appears in multiple nodes. Consider the following:
```csharp
void Main()
{
    var b = new B { X = "hi" };
    var a = new A { B1 = b, B2 = b};

}
class A
{
    public B B1 {get;set;}
    public B B2 {get;set;}
}
public class B
{
    public string X {get;set;}
}
```
If we ran it through JSON.NET, normally we would get this 
```json
{
  "B1": {
    "X": "hi"
  },
  "B2": {
    "X": "hi"
  }
}
```
But if we ran it with `PreserveReferencesHandling.Objects` setting turned on, we get this:
```json
{
  "$id": "1",
  "B1": {
    "$id": "2",
    "X": "hi"
  },
  "B2": {
    "$ref": "2"
  }
}
```
Which can become **much** more compact in larger trees (and is the only way to do it for ones that allow circular references)

**Reply by HaloFour** [reply]
> It can be breaking, but if it's implemented on an opt-in basis then the authors would have made that as a deliberate choice.

How would you suggest doing that?

**Reply by macsux** [reply]
I commented below on this already, but something like this synthesized by the compiler:
```csharp
protected virtual bool ShouldClone(string PropertyA, int PropertyB) => true;
```

**Reply by HaloFour** [reply]
IMO, it's not worth the complexity to records to add something like this, especially since it changes the public surface of the record and how consumers would interact with it.

**Reply by macsux** [reply]
How does it change public surface? It's a non breaking change unless explicitly opted in via record protected method override. 

Records have a lot of advantages over classes as far as concise syntax and non destructible mutations, but I already had to make a deliberate choice not to use them because of this limitation.

Serialization issue I mentioned is one of the major obstacles we've encountered so far on why custom built code generated was used instead of records.

**Reply by CyrusNajmabadi** [reply]
That solution would not work. Consider the following, library A references library B.  Library B updates and now adds this new property, and has it return false to indicate that a new copy should not be made. 

Library A still creates new copies because it is bypassing this API, breaking the semantics of the record in library B.


>How does it change public surface

It changes it because for it to work all downstream code has to be recompiled. If it wasn't changing the public surface, then you could make the change in B and it would just start working with all existing references without recompile.

**Reply by HaloFour** [reply]
> How does it change public surface?

That `ShouldClone` method would be a part of the public surface of the record, and the compiler would have to emit additional code in order to support it.

> Serialization issue I mentioned is one of the major obstacles we've encountered so far on why custom built code generated was used instead of records.

To me that sounds like a fault of the serialization library.  Records are supposed to have value identity, not reference identity.  But records are also not intended to be a shorthand replacement for wire protocol POCOs so if you want custom behavior code generation sounds like the appropriate solution.

**Reply by CyrusNajmabadi** [reply]
> protected virtual bool ShouldClone(string PropertyA, int PropertyB) => true;

I don't see how this helps. The *caller* is the one that clones. So they need to know if they can avoid that or not. So the caller needs a method to call. So it needs to be part of your public surface area.

**Reply by macsux** [reply]
>I don't see how this helps. The caller is the one that clones. So they need to know if they can avoid that or not. So the caller needs a method to call. So it needs to be part of your public surface area.

This would never be called by consumers of record directly, but is intended to be used by compiler, similar to the trick we play with a hidden `<Clone>$` method that only compiler has access to. 

> That solution would not work. Consider the following, library A references library B. Library B updates and now adds this new property, and has it return false to indicate that a new copy should not be made.


The choice of Library B **choosing** to enroll in this behaviour has a breaking change made by the library owner and it has affect on downstream consumers, so it should bump it's own major version because like you said, it is potentially breaking behaviour and may not be used as a safe drop-in replacement. While the public API surface didn't change, the change in behaviour of that record class is **explicit choice** of whoever owns that record, and they need to apply their own strategy to ensure downstream customers are aware of the change in the new version. 

>Library A still creates new copies because it is bypassing this API, breaking the semantics of the record in library B.

Why would it be bypassing it?  It has no choice in the matter as it would not be be called directly, it would instead be called by compiler similar to how we do with the hidden `<Clone>$` method

**Reply by CyrusNajmabadi** [reply]
>  While the public API surface didn't change

The public surface did change. That's why it's a breaking change. Existing code *must recompile* to pick it up and get the new behavior. That's a breaking change.

A non breaking change would be if B could be redeployed without recompiling downstream dependencies.

**Reply by CyrusNajmabadi** [reply]
> Why would it be bypassing it? 

Library A already shipped. Or it may be using an older compiler. It can only call your new methods (which must be public, not private) if it is recompiled.

**Reply by CyrusNajmabadi** [reply]
> similar to the trick we play with a hidden <Clone>$ method that only compiler has access to.

That doesn't help. It's still on the caller to call that new method.  So it's a *public* change. Your type is saying now *publicly* "call me with a different pattern than before". 

An internal change would have no effect on callers and you could redeploy your lib and get the new behavior, without anyone needing to care.

**Reply by macsux** [reply]
That's a fair point, I just didn't get what you meant by _public_ in this context. How about this. Atm the `with` clause on caller side is logically similar to this:
```csharp
var $temp = original.Clone();
$temp.MyProperty = "A";
var derived = $temp;
```

what if the internal implementation of `Clone` method (which is compiler generated) is the one that calls this new `ShouldClone`, and returns either a shallow clone or self. So for example, atm, compiler generates this:
```csharp
A <Clone>$ ()
{
    return new A (this);
}
```

after the change it would do this
```csharp
A <Clone>$ ()
{
    return ShouldClone() ? new A (this) : this;
}
```

**Reply by CyrusNajmabadi** [reply]
How does ShouldClone decide here?  It is passed in no data.

**Reply by CyrusNajmabadi** [reply]
Aside. Not wanting a new instance here effectively means you didn't want to treat this as a reference type. Have you considered just making this a struct record instead? Then there's no actual heao allocation occurring.

**Reply by macsux** [reply]
>How does ShouldClone decide here? It is passed in no data.

Yeah, sorry, brain cramp on that one. But how about this. If `ShouldClone(PropA, ProbB...)` is overriden in the record, the compiler generates a `<Clone>$(PropA, ProbB...)` instead of the parameter-less`<Clone>$()` method. This would make any existing downstream consumers **require** to use newer version of compiler (they can no longer compile against new version of this lib - they have to use newer compiler which will apply new _consumer side_ usage pattern) to be compatible with this new version of library that chose to use this new feature. 

So lets say lib B was v1.0 targeting `net8`. Then it chose to enroll in this behaviour, which hypothetically got introduced in .net 10. Since it's a breaking change, it bumped itself to v2.0 targeting `net10`. Library A that used to depend on B v1.0 cannot upgrade to v2.0 of this lib without also upgrading their project to target `net10` hence ensuring that they are using new compiler that can understand these new constructs.

> Have you considered just making this a struct record instead?

Yes - we need inheritance, and it wouldn't solve the issue with serializer optimization mentioned

**Reply by HaloFour** [reply]
> it wouldn't solve the issue with serializer optimization mentioned

I'd recommend taking that to the author of this serialization library.  There's nothing preventing them from using value identity to determine when to apply this optimization.

**Reply by macsux** [reply]
> I'd recommend taking that to the author of this serialization library. There's nothing preventing them from using value identity to determine when to apply this optimization.

How would a serializer know that it's a record type? Besides, reference check is cheap hence why optimization is offered only for classes (not structs) since equality check can be very expensive in context of serializers.

**Reply by HaloFour** [reply]
> How would a serializer know that it's a record type? 

They could detect the clone method, or require registration of the types that should be treated in this manner.  Sure, reference equality is cheaper, but that is the _one_ defining feature of record types is that they have value identity and not reference identity.

**Reply by HaloFour** [reply]
> `ShouldClone(PropA, ProbB...)`

This shape is problematic in that adding a new member to the record automatically changes the signature and breaks all existing consumers.  That is the reason that record cloning has the shape that it does, separating the cloning from the setting of the actual members.

**Reply by macsux** [reply]
> that is the _one_ defining feature of record types is that they have value identity and not reference identity.

That's not the only defining feature: records shipped with multiple features bundled into one "take it or leave it" construct. The `with` syntax is only available for records which allows implementing immutable and builder patterns easily. Immutability is not inherently tied to value equality - the way records shipped just chose to marry the two concepts. There might be legitimate reasons why I may want one without the other.

**Reply by macsux** [reply]
Yes, you're right. Conceptually, for this to work, we want to express intent on how an object will be mutated BEFORE it is cloned, via a variable number of mutation parameters, but without changing the contract. If you wrote this as user code in reflection, you could do it by passing in something like `Dictionary<Property,object>`, establishing a mapping between property and it's new intended value. Obviously we wouldn't do this at compiler or reflection, but I think it may be possible to do it IL level.  Conceptually would this work?:
- Caller assembles a stack allocated array of pointer pairs: first pointer being the property setter method, second one being the pointer to the new intended value.
- It passes this array to the `<Clone>$(array of pointer pairs)`. Since it's just one parameter, the contract stays the same even if new properties are added
- The library side Clone method would go over this array, which acts as a psuedo-dictionary of properties to new values. It could map this pseduo dictionary to positional parameters of the user provided `ShouldClone` method (since it knows that first argument corresponds to `this.PropertyA`, etc).
- It can then call user provided ShouldClone. 

This should theoretically avoid any allocations and contract stays stable as library is updated. Bit complex, but should work.
Thoughts?

**Reply by HaloFour** [reply]
I think it's very complex and would still impact the performance of all uses of records regardless of whether they would desire this niche functionality.  I also suspect that it would still be more expensive than the allocation would have been anyway.

**Reply by colejohnson66** [reply]
I agree with @HaloFour. The equality checks can get pretty expensive. Consider a string. After a certain string length, the cost of equality loses to just allocating a new object.

Not to mention that this request is such a niche thing. One of the main things records provide is the ability to *not* have to worry about reference equality. Deduping values on serialization should be handled by the serializer, not special language support. If your objects are large enough that deduping would be a meaningful improvement, then you might want to rethink your architecture.

**Reply by macsux** [reply]
> I think it's very complex and would still impact the performance of all uses of records regardless of whether they would desire this niche functionality. I also suspect that it would still be more expensive than the allocation would have been anyway.

I don't see how this would affect performance of users of all records since this is just an overload for those that chose to enrol in this behaviour. If you didn't override `ShouldClone`, then same compiler generated code is emitted like it is today (`Clone` method without parameters).

> Not to mention that this request is such a niche thing. One of the main things records provide is the ability to not have to worry about reference equality. Deduping values on serialization should be handled by the serializer, not special language support. If your objects are large enough that deduping would be a meaningful improvement, then you might want to rethink your architecture.

I feel like .NET have added A LOT of new features in recent years that most developers would find very neiche, low level and never use in their careers specifically to allow writing efficient code in libraries for advanced scenarios (which is often low level). Just some examples: `ValueTask`s, function pointers, `Span` /`Memory` objects, stack allocated arrays, `ref strut`s, just to name a few. As well, Microsoft has for years encouraged developers to overload `Equals` to provide more efficient implementations then what is provided out of the box. 

It's unfortunate that `with` syntax is limited to records as it provides clean way to add non-destructable mutations to your code, but have for some reason been joined at the hip to the "value equality" feature, whereas they are two completely distinct features/patterns.

**Reply by HaloFour** [reply]
> I don't see how this would affect performance of users of all records since this is just an overload for those that chose to enrol in this behaviour.

The virtual method has to always be there and has to always be invoked with the same contract otherwise a change to the internal implementation of the record would necessitate recompiling all projects that consume that record type.

> I feel like .NET have added A LOT of new features in recent years that most developers would find very neiche, low level and never use in their careers specifically to allow writing efficient code in libraries for advanced scenarios (which is often low level).

There have definitely been a number of lower level improvements to the language lately, but they've been driven by use cases that are widely applicable, often just because they automatically make the BCL much more efficient.  Many of those language changes can be directly tied to performance improvements in .NET over the last number of years.

**Reply by macsux** [reply]
> then you might want to rethink your architecture

The architecture is similar to Roslyn: immutable syntax language trees with visitor pattern to process them. Walking over the tree with a visitor can create two distinct versions of the tree on the heap, even if they are the same. Those trees can be huge when we're talking about representing codebase as an object node structure and can have serious implications to unnecessary memory allocations. I know Roslyn does a lot of tricks under the hood to try to optimize how it's immutability works.

**Reply by macsux** [reply]
BTW, look at behaviour of built in immutable structures in .NET:
```csharp
var a = ImmutableHashSet.Create(1);
var b = a.Add(1);
var c = a.Add(2);
Console.WriteLine(ReferenceEquals(a,b)); // true
Console.WriteLine(ReferenceEquals(a,c)); // false
```

`with` syntax is primarily used to drive immutable pattern, yet gives very little in terms of optimizing efficient implementations.

**Reply by macsux** [reply]
> The virtual method has to always be there and has to always be invoked with the same contract otherwise a change to the internal implementation of the record would necessitate recompiling all projects that consume that record type.

If owner of the record overridden the method, that is a breaking change of behaviour on the part of library owner. Even if "user code" contract stayed the same, the assembly is no longer serviceable and should not be substituted in favour of old one after compilation. Therefore, I think it's acceptable to require recompilation from downstream consumers if this happens. 

> There have definitely been a number of lower level improvements to the language lately, but they've been driven by use cases that are widely applicable

Immutability is a popular pattern though and at least in my experience is often considered for projects. However, every project I was on where the pattern was considered it inevitably devolved into debate over whether the tradeoff of what you get out of adopting this pattern is worth the
a) extensive boilerplate code the pattern brings (creating `With` mutation methods - which is why `with` syntax is so attractive as it's taken care of for you)
b) the cost of copying the object on each mutation is worth it and how much should be invested into optimizing efficiency of mutations to minimize the negative effect of copying

IMO this proposal would allow developers to at least solve for a. white giving them the necessary power to tackle b if the adoption of the pattern is contingent of necessary optimizations

**Reply by macsux** [reply]
BTW, since I seem to be getting a lot of pushback that `record`s are all about value equality, maybe this propsal is better suited to extending `with` syntax mutations to normal classes. Something like
`public immutable class A(P1 MyProperty);`

which would then apply the _non-destructive mutation_ behaviour we get on records to the normal class types without requiring "value equality" portion.

**Reply by HaloFour** [reply]
Even if .NET were to get something like "immutable classes" (which is a can of worms unto itself) that wouldn't be related to whether "withers" for such types would support this optimization to determine if the instance should be cloned or not.  The same problems would exist, in that every possible "wither" would have to incur the cost of such a comparison even if it was not necessary.  The value equality component is whether or not such an optimization should be necessary.  My opinion is that it's not, especially in the use case of serialization libraries, where the library could be modified to support value equality as well as reference equality through configuration policy.

**Reply by macsux** [reply]
Well like I mentioned, every "major" project that implements public surface immutability tends to try to optimize for memory allocations internally (I already gave examples of Roslyn, System.Collections.Immutable - and those optimizations are quite elaborate). So in my opinion is that it is something that's tackled when this pattern is applied.

Anyways, going back to original proposal, another idea that could potentially work that wouldn't require messing around with IL hacks is to generate a "hidden" buddy struct to act almost like a builder for data transfer. 

Would kinda looks like this from callers perspective:
```csharp
var builder = instance.<GetBuilder>$();
builder.Prop1 = newValue;
builder.Prop2 = otherValue;
var maybeCloned = builder.Build()
```

on the record side, how the builder is implemented will be determined by if `ShouldClone` is overriden:
### when not:
```csharp
public class A 
{
    public A(A copy) {}
    public string PropA {get;set;}

    public Builder <GetBuilder>$(A original) => new <Builder>$(this);
    public struct <Builder>$
    {
        A _clone;
        public <Builder>$(A original)
        {
            _clone = new A(original);
        }
        public string PropA
        {
            get => _clone.PropA;
            set => _clone.PropA = value;
        }
        public A Build() => _clone; // noop really, we already cloned it
    }
}
```

### when overridden:
```csharp
public class A 
{
    public A(A copy) {}
    public string PropA {get;set;}
    protected override bool ShouldClone(string propA)
    {
        // user logic
        return object.Equals(propA, propA);
    }

    public Builder <GetBuilder>$(A original) => new <Builder>$(this);
    public struct <Builder>$
    {
        A _original;
        
        public <Builder>$(A original)
        {
            _original = original;
        }
        public string PropA
        {
            get; set;
        }
        public A Build() => _original.ShouldClone(PropA) ? new A(_original) : _original;
    }
}
```

All data passing would happen only on the stack, contract would stay stable and overhead should be negligible, for non overridden scenario as it just acts as passthrough proxy to how current behaviour works.

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-03-23.md#builder-based-records

**Reply by macsux** [reply]
> See: https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-03-23.md#builder-based-records

Very nice. I assume since that was raised in 2020 it stalled and didn't go anywhere?
[I cooked up something very similar if you want to take a look while thinking of this problem to handle decision whether to clone or not](https://gist.github.com/macsux/6277d866b816222c8a36993a15cdb80b), though I especially like your approach to tracking set fields in a single member using bit flags

**Reply by HaloFour** [reply]
> Very nice. I assume since that was raised in 2020 it stalled and didn't go anywhere?

No, it didn't go anywhere because it was considered and rejected.

**Comment by huoyaoyuan** [comment]
> Since the two objects are immutable, returning original in place of copy should be safe.

Records are NOT guaranteed immutable:

```csharp
    public record Node(string MyProperty)
    {
        private int state;

        public int Change() => state++;
    }
```

And there's *no way to detect* if a record implementation is indeed immutable.

**Reply by macsux** [reply]
That's a fair point that they CAN be used implement immutability, but not guaranteed. How about something like this synthesized by the compiler:
```csharp
protected virtual bool ShouldClone(string PropertyA, int PropertyB) => true;
```

Default would always clone, but if overriden would allow opt-in behavior to potentially reuse existing instance.

**Reply by NN---** [reply]
What is the comparison strategy?
Just call `field.Equals(newvalue)` or something more complicated ?

**Reply by macsux** [reply]
Well, that would be up to the developer overriding this method. The compiler provided version of it just returns true so existing behaviour stays the same - we just adding a hook to allow it to be altered.
